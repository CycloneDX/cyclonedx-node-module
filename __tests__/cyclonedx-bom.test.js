"use strict";
const path = require("path");
const fs = require("fs");
const util = require("util");
const exec = util.promisify(require("child_process").exec);

const cliPath = path.resolve(__dirname, "../bin/cyclonedx-bom");

const emptyPackage = path.resolve(__dirname, "empty-package");
const simplePackage = path.resolve(__dirname, "simple-package");

const extBom1 = path.resolve(__dirname, "ext-bom.xml");
const extBom2 = path.resolve(__dirname, "ext-bom2.xml");

describe("cyclonedx-bom", () => {
  describe("BOM is valid", () => {
    test("package without dependencies", async () => {
      const { stdout, stderr } = await exec(`${cliPath} ${emptyPackage}`);
      expect(stderr).toEqual("");
      expect(stdout).toMatchSnapshot();
    });

    test("simple package", async () => {
      const { stdout, stderr } = await exec(`${cliPath} ${simplePackage}`);
      expect(stderr).toEqual("");
      expect(stdout).toMatchSnapshot();
    });

    test("simple package including devDependencies, -d", async () => {
      const { stdout, stderr } = await exec(`${cliPath} ${simplePackage} -d`);
      expect(stderr).toEqual("");
      expect(stdout).toMatchSnapshot();
    });

    test("simple package including devDependencies chane path and options order, -d", async () => {
      const { stdout, stderr } = await exec(`${cliPath} -d ${simplePackage}`);
      expect(stderr).toEqual("");
      expect(stdout).toMatchSnapshot();
    });
  });

  describe("help, -h", () => {
    test("-h prints help", async () => {
      const { stdout, stderr } = await exec(`${cliPath} -h`);
      expect(stdout).toMatchSnapshot();
      expect(stderr).toEqual("");
    });
  });

  describe("additional modules, -a", () => {
    test("-a adds modules from external BOM to empty bom", async () => {
      const { stdout, stderr } = await exec(
        `${cliPath} -a ${extBom1} ${emptyPackage}`
      );

      expect(stderr).toEqual("");
      expect(stdout).toMatchSnapshot();
    });

    test("-a adds modules from external BOM to simple bom", async () => {
      const { stdout, stderr } = await exec(
        `${cliPath} -a ${extBom1} ${simplePackage}`
      );

      expect(stderr).toEqual("");
      expect(stdout).toMatchSnapshot();
    });

    test("-a adds modules from external BOM to simple bom with devDependencies", async () => {
      const { stdout, stderr } = await exec(
        `${cliPath} -d -a ${extBom1} ${simplePackage}`
      );

      expect(stderr).toEqual("");
      expect(stdout).toMatchSnapshot();
    });

    test("multiple additional BOMs", async () => {
      const { stdout, stderr } = await exec(
        `${cliPath} -d -a ${extBom1} -a ${extBom2} ${path.resolve(
          __dirname,
          "simple-package"
        )}`
      );

      expect(stderr).toEqual("");
      expect(stdout).toMatchSnapshot();
    });
  });
});
