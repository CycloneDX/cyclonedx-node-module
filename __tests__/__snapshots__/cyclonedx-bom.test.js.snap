// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cyclonedx-bom BOM is valid package without dependencies 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>
<bom xmlns=\\"http://cyclonedx.org/schema/bom/1.0\\" version=\\"1\\">
    <components>
    </components>
</bom>
"
`;

exports[`cyclonedx-bom BOM is valid simple package 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>
<bom xmlns=\\"http://cyclonedx.org/schema/bom/1.0\\" version=\\"1\\">
    <components>
        <component type=\\"library\\">
            <group>scope</group>
            <name>scoped-package</name>
            <version>1.0.0</version>
            <description>
                <![CDATA[]]>
            </description>
            <licenses>
                <license>
                    <id>ISC</id>
                </license>
            </licenses>
            <purl>pkg:npm/%40scope/scoped-package@1.0.0</purl>
            <modified>false</modified>
        </component>
    </components>
</bom>
"
`;

exports[`cyclonedx-bom BOM is valid simple package including devDependencies chane path and options order, -d 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>
<bom xmlns=\\"http://cyclonedx.org/schema/bom/1.0\\" version=\\"1\\">
    <components>
        <component type=\\"library\\">
            <group>scope</group>
            <name>scoped-package</name>
            <version>1.0.0</version>
            <description>
                <![CDATA[]]>
            </description>
            <licenses>
                <license>
                    <id>ISC</id>
                </license>
            </licenses>
            <purl>pkg:npm/%40scope/scoped-package@1.0.0</purl>
            <modified>false</modified>
        </component>
        <component type=\\"library\\">
            <name>empty-package</name>
            <version>1.0.0</version>
            <description>
                <![CDATA[]]>
            </description>
            <licenses>
                <license>
                    <id>ISC</id>
                </license>
            </licenses>
            <purl>pkg:npm/empty-package@1.0.0</purl>
            <modified>false</modified>
        </component>
    </components>
</bom>
"
`;

exports[`cyclonedx-bom BOM is valid simple package including devDependencies, -d 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>
<bom xmlns=\\"http://cyclonedx.org/schema/bom/1.0\\" version=\\"1\\">
    <components>
        <component type=\\"library\\">
            <group>scope</group>
            <name>scoped-package</name>
            <version>1.0.0</version>
            <description>
                <![CDATA[]]>
            </description>
            <licenses>
                <license>
                    <id>ISC</id>
                </license>
            </licenses>
            <purl>pkg:npm/%40scope/scoped-package@1.0.0</purl>
            <modified>false</modified>
        </component>
        <component type=\\"library\\">
            <name>empty-package</name>
            <version>1.0.0</version>
            <description>
                <![CDATA[]]>
            </description>
            <licenses>
                <license>
                    <id>ISC</id>
                </license>
            </licenses>
            <purl>pkg:npm/empty-package@1.0.0</purl>
            <modified>false</modified>
        </component>
    </components>
</bom>
"
`;

exports[`cyclonedx-bom additional modules, -a -a adds modules from external BOM to empty bom 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>
<bom xmlns=\\"http://cyclonedx.org/schema/bom/1.0\\" version=\\"1\\">
    <components>
        <component type=\\"library\\">
            <group>external-bom</group>
            <name>package</name>
            <version>1.0.0</version>
            <description>
                <![CDATA[]]>
            </description>
            <licenses>
                <license>
                    <id>ISC</id>
                </license>
            </licenses>
            <purl>pkg:npm/%40external-bom/package@1.0.0</purl>
            <modified>false</modified>
        </component>
    </components>
</bom>
"
`;

exports[`cyclonedx-bom additional modules, -a -a adds modules from external BOM to simple bom 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>
<bom xmlns=\\"http://cyclonedx.org/schema/bom/1.0\\" version=\\"1\\">
    <components>
        <component type=\\"library\\">
            <group>scope</group>
            <name>scoped-package</name>
            <version>1.0.0</version>
            <description>
                <![CDATA[]]>
            </description>
            <licenses>
                <license>
                    <id>ISC</id>
                </license>
            </licenses>
            <purl>pkg:npm/%40scope/scoped-package@1.0.0</purl>
            <modified>false</modified>
        </component>
        <component type=\\"library\\">
            <group>external-bom</group>
            <name>package</name>
            <version>1.0.0</version>
            <description>
                <![CDATA[]]>
            </description>
            <licenses>
                <license>
                    <id>ISC</id>
                </license>
            </licenses>
            <purl>pkg:npm/%40external-bom/package@1.0.0</purl>
            <modified>false</modified>
        </component>
    </components>
</bom>
"
`;

exports[`cyclonedx-bom additional modules, -a -a adds modules from external BOM to simple bom with devDependencies 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>
<bom xmlns=\\"http://cyclonedx.org/schema/bom/1.0\\" version=\\"1\\">
    <components>
        <component type=\\"library\\">
            <group>scope</group>
            <name>scoped-package</name>
            <version>1.0.0</version>
            <description>
                <![CDATA[]]>
            </description>
            <licenses>
                <license>
                    <id>ISC</id>
                </license>
            </licenses>
            <purl>pkg:npm/%40scope/scoped-package@1.0.0</purl>
            <modified>false</modified>
        </component>
        <component type=\\"library\\">
            <name>empty-package</name>
            <version>1.0.0</version>
            <description>
                <![CDATA[]]>
            </description>
            <licenses>
                <license>
                    <id>ISC</id>
                </license>
            </licenses>
            <purl>pkg:npm/empty-package@1.0.0</purl>
            <modified>false</modified>
        </component>
        <component type=\\"library\\">
            <group>external-bom</group>
            <name>package</name>
            <version>1.0.0</version>
            <description>
                <![CDATA[]]>
            </description>
            <licenses>
                <license>
                    <id>ISC</id>
                </license>
            </licenses>
            <purl>pkg:npm/%40external-bom/package@1.0.0</purl>
            <modified>false</modified>
        </component>
    </components>
</bom>
"
`;

exports[`cyclonedx-bom additional modules, -a multiple additional BOMs 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>
<bom xmlns=\\"http://cyclonedx.org/schema/bom/1.0\\" version=\\"1\\">
    <components>
        <component type=\\"library\\">
            <group>scope</group>
            <name>scoped-package</name>
            <version>1.0.0</version>
            <description>
                <![CDATA[]]>
            </description>
            <licenses>
                <license>
                    <id>ISC</id>
                </license>
            </licenses>
            <purl>pkg:npm/%40scope/scoped-package@1.0.0</purl>
            <modified>false</modified>
        </component>
        <component type=\\"library\\">
            <name>empty-package</name>
            <version>1.0.0</version>
            <description>
                <![CDATA[]]>
            </description>
            <licenses>
                <license>
                    <id>ISC</id>
                </license>
            </licenses>
            <purl>pkg:npm/empty-package@1.0.0</purl>
            <modified>false</modified>
        </component>
        <component type=\\"library\\">
            <group>external-bom</group>
            <name>package</name>
            <version>1.0.0</version>
            <description>
                <![CDATA[]]>
            </description>
            <licenses>
                <license>
                    <id>ISC</id>
                </license>
            </licenses>
            <purl>pkg:npm/%40external-bom/package@1.0.0</purl>
            <modified>false</modified>
        </component>
        <component type=\\"library\\">
            <group>external-bom</group>
            <name>package2</name>
            <version>1.0.0</version>
            <description>
                <![CDATA[]]>
            </description>
            <licenses>
                <license>
                    <id>ISC</id>
                </license>
            </licenses>
            <purl>pkg:npm/%40external-bom/package2@1.0.0</purl>
            <modified>false</modified>
        </component>
    </components>
</bom>
"
`;

exports[`cyclonedx-bom help, -h -h prints help 1`] = `
"Usage:  cyclonedx-bom [OPTIONS] [path]
Options:
  -h        - this help
  -a <path> - merge in additional modules from other scanner
  -o <path> - write to file instead of stdout
  -d        - include devDependencies
"
`;
